shared/p2p/service_test.go

```diff
@@ -70,15 +71,101 @@ func TestP2pPortTakenError(t *testing.T) {
 }
 
 func TestBroadcast(t *testing.T) {
-	s, err := NewServer(&ServerConfig{})
+	servers, err := runP2PServersWithDifferentPorts(3)
 	if err != nil {
-		t.Fatalf("Could not start a new server: %v", err)
+		t.Fatalf("error while trying to create server: %s", err)
+	}
+
+	err = connectServersTo(servers[0], servers[1:])
+	if err != nil {
+		t.Fatalf("error while trying to connect to peer: %s", err)
 	}
 
 	msg := &shardpb.CollationBodyRequest{}
-	s.Broadcast(msg)
+	subscribeServersToTopic(servers, "theTopic", msg)
+
+	time.Sleep(20 * time.Millisecond)
+
+	msgSubsChannel := make(chan Message)
+	for _, server := range servers[1:] {
+		server.Subscribe(&shardpb.CollationBodyRequest{}, msgSubsChannel)
+	}
+
+	aMessage := &shardpb.CollationBodyRequest{ShardId: 1234}
+	servers[0].Broadcast(aMessage)
+
+	doneChan := make(chan bool)
+	errorChan := make(chan bool)
+	timeoutChan := make(chan bool)
+
+	var wg sync.WaitGroup
+	wg.Add(len(servers[1:])) // Num of nodes that receive the channel
+
+	go func() {
+		for recMessage := range msgSubsChannel {
+			protoMsg := recMessage.Data.(*shardpb.CollationBodyRequest)
+			if protoMsg.ShardId != aMessage.ShardId {
+				errorChan <- true
+			}
+
+			wg.Done()
+		}
+	}()
+
+	go func() {
```

Comment:
Please add a comment to explain each of these go routines. It took me a bit of thinking to understand what is going on here.

Reply:
Ok no problem!